[HVG-588] Feature / proper sign status messages

styrelius opened the pull request:
## What?

<!-- What changes are made? If there are many significant changes, a list might be a good format. -->
Add better sign status messages rendered in `SignStatus` component, including a generic error message when something goes wrong when initiating the signing (i.e. you've already signed or something else is off):
1) New text key `CHECKOUT_SIGN_GENERIC_ERROR` when is `signUiState` is `'FAILED'`.
2) New `LinkToChat` component which renders a `TextButton` that opens Intercom chat when clicked. This is rendered with the generic error message.
3) A message when `signUiState` is `'STARTED_WITH_REDIRECT'` (which s true when you click sign in Norway and Denmark).
4) Some refactoring of `Checkout`, `Sign` and `Sign` components to make it clearer _to me_ what's going on there. 
**ADDITION:** New file `CheckoutSuccessRedirect.tsx` where I moved the Swedish success redirect and the `<TrackAction>` wrapping it.
5) New translations of some of the text keys used for Bank ID statuses. ~~A bit unclear to me though when some of these statuses are even returned/used.~~ 
**EDIT:** Also new names for the text keys so that it's clearer that they are only for _Swedish_  BankID, and  (after talking to backender Fredrik) I removed some of them that I deemed unnecessary.
6) Trigger the hiding and showing of the Intercom "launcher" (button) from `Checkout` component instead of from `OfferNew` component - this way the launcher is hidden even if you refresh the `/sign` page.
7) Clarify with the sign button text in the Swedish checkout that you have to sign with _mobile_  BankID.

_Please note_ that the possibility to cancel the signing session that is referenced in the ticked was too complicated to add at this point, since there is no support for it in the backend and thus we would have to make some fake cancelling or whatever.



## Why?

<!-- Why are these changes made? -->

1) The main problem here has been that if the signing never starts due to some issues on the backend the user is not notified in any way - the sign button just shows a quick spinner and is then back to normal.
2) Another smaller issue is that if you are redirected to another url to sign (which you are in Norway and will be in Denmark) and this takes some time you won't get any information about what's going on.
3) Also some of the other status messages for the Swedish signing have implied that you could use BankID on your computer or on card to sign, which you really shouldn't do since you would not be able to log in to the Hedvig app without a mobile BankID on the device where you have the app installed. Overall we need to be clearer that this is _mobile_  BankID only for Sweden.


<!-- If there is one, add a link to the Jira ticket below. -->
_Referenced ticket [HVG-588]_


<!-- And, if that makes sense, add screenshots and/or GIF's below -->
## GIF's
**Swedish sign, successful after user cancelling sign in BankID app and then starting sign again:**
![2020-12-18-SE-sign-cancelled-and-successful](https://user-images.githubusercontent.com/42962286/102597372-d4cb9900-411a-11eb-8afa-d2bbe68bacbe.gif)

**Swedish sign, failed:**
![2020-12-14-SE-sign-failed](https://user-images.githubusercontent.com/42962286/102139370-0d8b1a00-3e5e-11eb-8bf5-920704c4d38e.gif)

**Norwegian sign, successful:**
![2020-12-14-NO-sign-successful](https://user-images.githubusercontent.com/42962286/102139442-2398da80-3e5e-11eb-8e76-2be56395be7b.gif)

**Norwegian sign, failed:**
![2020-12-14-NO-sign-failed](https://user-images.githubusercontent.com/42962286/102139467-2abfe880-3e5e-11eb-96eb-6a52a0190d74.gif)



[HVG-588]: https://hedvig.atlassian.net/browse/HVG-588

styrelius committed:
Change the SignUiState type in Sign.tsx from an enum to a union type since that makes it easier for me to read the code. Adjust Checkout component to this change.

styrelius committed:
Refactor SignStatus.tsx with a useEffect instead of having a lot of if statements baked into the JSX.

styrelius committed:
Add useEffect in SignStatus component to set signStatusTextKey to new text key CHECKOUT_SIGN_GENERIC_ERROR if signUiState is 'FAILED'.

styrelius committed:
Refactor the 'startSign' function in Checkout component to use a promise instead of async/await since I thought it made more sense here - this also enabled to add a 'catch' where signUiState is set to 'FAILED', in a simple way.

styrelius committed:
Fix onClick function in Sign.tsx - remove the 'async' since it's unnecessary and remove the 'if' statement making a return when 'canInitiateSign' is false, since the button is disabled with that exact condition.

styrelius committed:
Fix some status messages in SignStatus.tsx:
Remove the special sign status msg for signState 'initiated' since that's not really useful.
Add sign status msg if the signUiState is 'STARTED_WITH_REDIRECT' so the user can be told they are being redirected.

styrelius committed:
Trigger hiding and showing of Intercom launcher in Checkout component instead of from OfferNew component. This makes the Intercom launcher hidden even after you refresh the page when Checkout sidebar is open.

styrelius committed:
Add new component LinkToChat, with a 'TextButton' that opens Intercom chat. Import and conditionally render it in SignStatus.tsx, depending on new state 'isChatLinkVisible' (set to true if the signin fails in _any way_).

styrelius committed:
Merge branch 'master' into feature/proper-sign-status

styrelius committed:
Download translations of CHECKOUT_SIGN_GENERIC_ERROR and CHECKOUT_SIGN_LINK_TO_OPEN_CHAT.

styrelius committed:
Download Lokalise translations for CHECKOUT_SIGN_STARTED_WITH_REDIRECT textkey.

styrelius committed:
Rename isChatLinkVisible state, to hasSigningError, in SignStatus.tsx. Add a li'l WarningIcon by the text if that state is true.

styrelius committed:
Download translations with some fixes of sign status messages.

styrelius committed:
Download Lokalise translations again since I made some changes there.

gustaveen reviewed (commented):
Good job Ellen, really nice with the chat link if the sign fails! 👌 Will test a bit locally to understand which errors are triggered

styrelius commented on src/client/pages/OfferNew/Checkout/index.tsx:
```diff
@@ -225,11 +220,12 @@ export const Checkout: React.FC<Props> = ({
   const canInitiateSign = Boolean(
-    signUiState !== SignUiState.STARTED &&
-      signUiState !== SignUiState.STARTED_WITH_REDIRECT &&
+    signUiState !== 'STARTED' &&
+      signUiState !== 'STARTED_WITH_REDIRECT' &&
```
I really disagree actually - I think it's so much easier to read with this kind of string union type. I changed it because I found it so hard to read the code with the enum we previously had. But if everyone else likes enums more I can of course change it.

gustaveen replied:
I like unions as well 👌, but I'm guessing there's a reason that we have enums in most cases like this. Since I'm no typescript wiz, what's you take @palmenhq @virrios

styrelius replied:
I asked @palmenhq about that before, like more generally the difference between using enums and unions like these, and the way I understood it from his answers and from how enums works in these kinds of cases, they are completely interchangeable.

virrios replied:
Fun tidbit since, we did this last week, (thought it was not necessary once we removed all danish type of Contracts and got removed)
Enums are iterable.  It was not at first possible for types but was later introduced, with this syntax
`
const contracts = ['travel', 'home', 'accident'] **as const;**
type IterableContractsType = typeof contracts[number]; // 'travel' | 'home' | 'accident'
`
I don't mind the enum, in terms of readability, but now its not really typesafe since its a numeric enum? And its actually passed as props in some components. 
I don't mind the readability of the union type either, but comparing when you are using it as to compare states I feel like its more natural to have it as a enum. 

I am not that strongly opinionated in this though 😄

styrelius replied:
> I don't mind the enum, in terms of readability, but now its not really typesafe since its a numeric enum? And its actually passed as props in some components.

I'm not sure I understand fully @virrios - what is not really typesafe?

But yes, it's true that enums are iterable but in this case we're not using that (or any other of the special properties of the enum, as far as I can see) so here it's perfectly possible to use a union with strings. What I like is the readability when referencing those values, which in my view comes from the enum reference being more cluttered:
```javascript
  // less clutter :)
  signUiState !== 'STARTED'

  // more clutter
  signUiState !== SignUiState.STARTED
```
Especially confusing in these particular files where everything is called something with "sign" and "state" or "status" 😄

virrios replied:
> I'm not sure I understand fully @virrios - what is not really typesafe?
> 
> But yes, it's true that enums are iterable but in this case we're not using that (or any other of the special properties of the enum, as far as I can see) so here it's perfectly possible to use a union with strings. What I like is the readability when referencing those values, which in my view comes from the enum reference being more cluttered:

Sry what I meant regarding that the enums are iterable was just to reference last week when we discussed the **strange** `as const` or `<const>['....']` syntax and I thought it was interesting and **fun** information **but not really applicable to what we are doing here 😄** 

**Regarding typesafeness** 
**This is numeric Enum because its auto incremented** 
`
enum Example {
a,
b
}
`
Its not type safe in the sense that If you pass it to a function, you can pass an arbitrary number as well. This type of enum is useful to use as a mask. However,  I saw that we use it as a property in another component and I think that you can pass just a number and it would not cause an error?. **This is my understanding** of numeric enums and  **correct me if I am wrong here!**.

So the union type would be more typesafe (or make the enum a string enum). 
So this was more to add to the general discussion.

** Regarding readability ***
I like enums and I think they are readable and not necessarily more cluttered.


Maybe we should decide when/when not to use enums/union? I like what you said in the meeting, so it looks like one person is writing the code.

styrelius replied:
@virrios Ah, now I see! Interesting, didn't really know that about enums. But yes, let's have that discussion elsewhere then, and leave this with the changes I made since those are more type safe as Victor said?

palmenhq replied:
Sorry i'm a bit late to the party here... Interesting point about the enums vs numbers, i had no idea but as it seems typescript is happy with the following code? Seems bad to me. I guess it's quite obvious that you shouldn't interchange enums with numbers but you really have a point here. For this case i think union types is good.

```ts
enum MyEnum { A, B }
const fn = (blah: MyEnum) => blah

fn(MyEnum.A) // OK
fn(0) // OK
fn(12) // OK??? like wat
```

styrelius commented on src/translations/da_DK.json:
```diff
@@ -239,5 +239,12 @@
   "SSN_DOES_NOT_MATCH_BIRTH_DATE": "",
-  "INVALID_SSN_LENGTH": ""
+  "INVALID_SSN_LENGTH": "",
+  "CHECKOUT_SIGN_GENERIC_ERROR": "Noget gik galt. Prøv igen eller kontakt os.",
+  "CHECKOUT_SIGN_LINK_TO_OPEN_CHAT": "Chat med os",
```
This is the Danish translation, the Norwegian looks exactly like your suggestion 🙃

gustaveen replied:
Lol, I blame my bad morning brain 😂 Sorry for that! 🤦‍♂️

styrelius replied:
Which I guess shows how skilled you are at Norwegian 🥇

gustaveen replied:
Haha, or maybe not 😂

gustaveen commented on src/client/pages/OfferNew/Checkout/index.tsx:
```diff
@@ -201,6 +201,15 @@ export const Checkout: React.FC<Props> = ({
+      Intercom('update', { hide_default_launcher: false })
+    }
+  }, [visibilityState])
+
```
Awesome! 🙌

styrelius replied:
😊 And this could easily be made into a custom hook if we want to hide the Intercom launcher somewhere else in the future.

gustaveen commented on src/client/pages/OfferNew/Checkout/index.tsx:
```diff
@@ -231,33 +231,34 @@ export const Checkout: React.FC<Props> = ({
 
     const baseUrl = `${window.location.origin}/${locale}/new-member`
-    const result = await signQuotes({
+    signQuotes({
```
Just my opinion, I prefer `async/await` 🤓

styrelius replied:
Haha OK I totally respect that! 😄  The thing with using a `Promise` here was 1) that I could do a destructuring of `data` in a (to me) neat way and 2) that it enabled to easily add a `catch` without wrapping it all in a `try/catch` or having yet _another_ `if` to check if the request failed.

virrios replied:
I have seen the usage of  the "then" to handle promises in our code but also await. I think we should be consistent
and the await syntax is more modern and a bit cleaner especially if have several steps.

styrelius replied:
I feel that both `Promises` and `async/await` they can both be useful for different cases and I'm not really sure that this is a case where you have to pick one? Although I generally love consistency 😀

palmenhq replied:
Just to chip in a third opinion i prefer `pipe` from rxjs ;)

Jokes aside i don't really have an opinion either. If we want to go towards a more functional style code base then maybe we should consider using `then` more (as generally it results in chaining functions rather than assigning tons of variables and hence state) but otherwise i agree `async`/`await` is a bit more readable. For this case imo it was nicer with `async`/`await` to be honest, but also we didn't actually catch if there was any errors. Adding a catch around this with nested ifs would decrease the prettiness quite a bit though, so i'd vote for keeping it with `then`s 

**tldr: keep it with `then` imo**

virrios commented on src/client/pages/OfferNew/Checkout/SignStatus.tsx:
```diff
@@ -29,47 +49,58 @@ type Props = {
+            )}
+            {textKeys[signStatusTextKey]()}
+          </StatusText>
+          {hasSigningError && <LinkToChat />}
```
Awesome feature!fT! 🚀 🎆

styrelius committed:
Reset 'hasSigningError' if 'signUiState' is no longer 'FAILED'.

styrelius committed:
Move the Redirect and TrackAction component wrapping it (that should be rendered if 'signStatus.signState' is 'COMPLETED') to a new file: CheckoutSuccessRedirect.tsx.

styrelius committed:
Fix issue with offer page not being rendered due to Intercom not being defined initially in /Checkout/index.tsx:
Change the condition for ceasing to hide the Intercom launcher to visibilityState being 'CLOSING' instead of 'CLOSED' ('CLOSED' is the initial state which means the code in that block would be executed immediately, causing error).
Also move UseScrollLock to a more logical place in the file.

styrelius committed:
Step up status message game:
Clean up among status messages based on BankID status codes - all of them weren't necessary.
Set signUiState to 'FAILED' in Checkout component if the BankID status is 'failed'.
Fix a bunch of conditionals in useEffect in SignStatus.tsx to get the messages and the WarningIcon and LinkToChat rendered at a correct time. This includes passing isLoading state from Sign component down to SignStatus.

styrelius committed:
Move all the messy annoying logic deciding which value to set as statusTextKey from the useEffect in SignStatus.tsx to a new function outside of the component.

styrelius committed:
Dowload translations to get new sign text keys (and some other as well apparently). Plus fix text key that wasn't named properly in SignStatus.tsx.

styrelius committed:
Change 'StatusMessageTextKey' from a union type to an enum.

palmenhq commented on src/client/pages/OfferNew/Checkout/SignStatus.tsx:
```diff
@@ -1,75 +1,132 @@
+
+enum StatusMessageTextKey {
+  REDIRECTING = 'CHECKOUT_SIGN_STARTED_WITH_REDIRECT',
+  GENERIC_ERR = 'CHECKOUT_SIGN_GENERIC_ERROR',
+  START_SIGNING = 'SE_BANKID_START_SIGN',
+  CERTIFICATE_ERR = 'SE_BANKID_CERTIFICATE_ERROR',
```
this is nit but i'm not sure i like shortening it to `err`. Err is i guess quite well accepted but i see no real reason here so we could just write it out imo

styrelius replied:
I agree, I was just annoyed when selecting all instances of `CERTIFICATE_ERROR` I also got that part of the `SE_BANKID_CERTIFICATE_ERROR` 😄 So I just wanted to differentiate them in some way, but I actually don't like `err` either. Can change it!

palmenhq reviewed (dismissed):
![](https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fwww.reactiongifs.com%2Fr%2F2012%2F06%2Fself_high_five.gif&f=1&nofb=1)

THIS. IS. GREAT!! I had two more comments but you addressed them before i submitted my review haha. Very good job with this!

styrelius committed:
Replace 'ERR' with 'ERROR' in SignStatus.tsx.

styrelius committed:
Merge branch 'master' into feature/proper-sign-status

virrios reviewed (dismissed):
Awesome job, and a super neat feature, for sure improve the user experience when signing!

gustaveen commented:
Boom! Well done @styrelius 🍾 🎉

styrelius committed:
Download translations again, hopefully this will solve merge conflict with master.

styrelius committed:
Merge branch 'master' into feature/proper-sign-status. Fix conflicts in translation files.

gustaveen reviewed (approved):
🚢
